File: stack-linked-list/stack_linked_list.cpp
/*********
 * Stack using linked list
 * 
 * UNIVESP - www.univesp.br
 * Computer Engineering
 * Data structures discipline
 */

#include "stack.h"
#include <cstddef>  // to work NULL
#include <new>

#include <iostream>
using namespace std;

Stack::Stack()
{
    top = NULL;
}

Stack::~Stack()
{
    NodeType* tempPtr;
    while (top != NULL)
    {
        tempPtr = top;
        top = top -> next;
        delete tempPtr;
    }
}

bool Stack::isEmpty() const
{
    return (top == NULL);
}

bool Stack::isFull() const
{
    NodeType* location;
    try
    {
        location = new NodeType;    
        delete location;            // instantiated, memory is not full!
        return false;

    } catch (bad_alloc exception)   
    {                               // fail, memory is full!
        return true;
    }
}

void Stack::push(ItemType item)
{
    if (isFull())
    {
        throw "Stack is already full!";
    }

    NodeType* location;
    location = new NodeType;
    location->info = item;
    location->next = top;
    top = location;
}

ItemType Stack::pop()
{
    if (isEmpty())
    {
        throw "Stack is empty!";
    }

    NodeType* tempPtr;
    tempPtr = top;
    ItemType item = top->info;
    top = top->next;
    delete tempPtr;
    return item;
}

void Stack::print() const
{

    NodeType* tempPtr = top;
    while (tempPtr != NULL)
    {
        cout << tempPtr->info;
        tempPtr = tempPtr->next;
    }
    cout << endl;
}

File: Stack-Java/StackLinkedList.java
/*********
 * Stack using linked list
 * 
 * UNIVESP - www.univesp.br
 * Computer Engineering
 * Data structures discipline
 */

package com.crv;

public class StackLinkedList<T> implements Stack<T> {

    private Node<T> top = null;

    private static class Node<T> {
        public T item;
        public Node next;
    }

    @Override
    public boolean isEmpty() {
        return top == null;
    }

    @Override
    public void push(T item) {
        Node<T> newNode = new Node<>();
        newNode.item = item;
        newNode.next = top;
        top = newNode;
    }

    @Override
    public T pop() {
        if (isEmpty())
            throw new RuntimeException("Stack is empty!");

        T aux = top.item;
        top = top.next;

        return aux;
    }

}
